accepted/future-releases/0546-patterns/feature-specification.md

```diff
@@ -1885,18 +1885,23 @@ var [int a, b] = <num>[1, 2];   // List<num>.
 To type check a pattern `p` being matched against a value of type `M`:
 
 *   **Logical-or** and **logical-and**: Type check each branch using `M` as the
-    matched value type.
+    matched value type. The required type of the pattern is `Object?`.
```

Comment:
I think you mean `int x && < 4`, but otherwise yes.

Reply:
Right! I thought at first that the required type would only be used in an irrefutable context, but that's not true, the same concept is used for the dynamic semantics of matching. So it _is_ possible for that `int x` to be a type test that promotes the scrutinee.

We need to make a distinction between `||` and `&&`: For `||` we may proceed to match the right operand in the case where the first type test implied by a successful match in the left operand failed, so the right operand needs to use `M` as the matched value type. For `&&` we know that we're in the true continuation of the left operand when the right operand is being matched, so that one can rely on the promotions in the left operand.

Reply:
Generally, we can treat the matched value as being a fresh promotable variable, and have test promote it as usual.

```dart
Object? f() =>;
switch (f()) {
   case int(sign: >= 0) && < 4: ...   // synthetic variable _1 bund to `f()` promoted to `int`.
}
```

The big question is what happens when we switch on an existing variable. Say:
```dart
Object? o = ...;
switch (o) {
   case int y && < 4: ...
   case int(sign: >=) when o < 5: ... // valid? Is `o` promoted to `int` as well as the matched value.
}
```

We can't say that we always use `o` as the variable, because we cache the matched value and the variable can change. That's visible from:
```dart
Object? o = 1;
switch (o) {
  case int y when o = false:  throw "Nope";
  case int z: // z is `int`, `o` is not.
}
```
We can choose to promote as long as we can prove that `o` is unchanged, but if it changes type, we can no longer assume that tests on the matched value also reflect tests on the original variable.

Or we can choose to not do so, but that's also annoying. I don't like it when `if` is more powerful than `switch`, because it means we are dumbing down the `switch` relative to what we already consider reasonable deductions. (But it might be reasonable, because there are differences, like doing multiple tests on a cached value, which a plain `if` won't do.)



About `||` and `&&`, how promotions flow is nothing new. Both `&&` and `||` can propagate promotion of the matched value from the first pattern into the second. Example:
```dart
int? o = ...;
switch (o) { 
  case == null || >= 0: ...
```
The failure continuation for `<int? value> == null` promotes the value to non-nullable, as normal.

A test can have a possible promotion on both success and failure. The `FutureOr<int> x = ..; if (x is Future<int>) { ... promoted to Future<X> ... } else { ... promoted to int ... }` test is the most visible example.